Fixing infinite loop when merging from a rename changeset

jnm2 opened the pull request:
Please do not merge until the fix is finished. (Will fix #873)

Currently when git-tfs tries to clone a merge changeset and a merge parent is itself a rename changeset, git-tfs loops forever. It looks similar to this:

```
Tfs branches found:
- $/A
- $/B
=> Working on TFS branch : $/A
=> Working on TFS branch : $/B
=> Working on TFS branch : $/Trunk
        Fetching from dependent TFS remote B'...
=> Working on TFS branch : $/Trunk
        Fetching from dependent TFS remote B'...
=> Working on TFS branch : $/Trunk
        Fetching from dependent TFS remote B'...
=> Working on TFS branch : $/Trunk
        Fetching from dependent TFS remote B'...
=> Working on TFS branch : $/Trunk
        Fetching from dependent TFS remote B'...
=> Working on TFS branch : $/Trunk
        Fetching from dependent TFS remote B'...
[...endless]
```

@spraints @pmiossec @ivan-danilov I want to get your comments on this. There seem to be several ways to fix this so that all the tests pass, but I'm not sure which way is the right approach.

@elangelo's PR [#956](https://github.com/git-tfs/git-tfs/pull/956/files?w=1) doesn't come with any tests but it does make my test pass. I'm having doubts it is the right approach, especially [this line](https://github.com/git-tfs/git-tfs/pull/956/files?w=1#diff-bc06152b842b2133c8a7eefb9d446644R63).

@icnocop's PR #927 was also supposed to address this issue, but at least for my scenario, it had no effect.

jnm2 commented:
I believe I have a fix that begins to makes sense (to an outsider).

jnm2 commented:
Appveyor failed but all tests pass on my machine. I am currently trying to figure out how to run tests identically to Appveyor.

jnm2 commented:
Yeah, not sure what to say. Did git clean -xfd, built a release, ran tests with the same command lines as appveyor.

jnm2 commented:
Ah, AppVeyor runs an additional `.\FunctionalTesting\smoke_tests.ps1 GitTfs\bin\Release`. I'll need to see why that's failing.

jnm2 commented:
@spraints @pmiossec @ivan-danilov Can someone verify whether this means the test is wrong or the implementation is wrong? https://ci.appveyor.com/project/pmiossec/git-tfs-v2qcm/build/0.25.1.539#L609

spraints commented on GitTfsTest/TestHelpers/AssertTimeout.cs:
```diff
@@ -0,0 +1,36 @@
+
+            if (Debugger.IsAttached)
+            {
+                thread.Join();
```
I don't remember what c# does with exceptions on a child thread. Will an exception raised by the child get thrown here, too? The [docs](https://msdn.microsoft.com/en-us/library/95hbf2ta.aspx) don't say, so my guess is that they don't get thrown by `Join` either. Should this re-throw errors from the child thread?

jnm2 replied:
@spraints Good catch! I overlooked that. Every threadstart should be wrapped in `try...catch` because otherwise instead of the caller being able to handle it, the entire process goes down.

jnm2 replied:
@spraints Fyi, you have `<LangVersion>5</LangVersion>` set only for Debug configuration. That confused me for a bit. Ideally <`LangVersion>` is set in an unconditional propertygroup.

jnm2 replied:
(The VS tooling won't get you there though.)

spraints replied:
> Ideally `<LangVersion>` is set in an unconditional propertygroup.

Feel free to make this change.

jnm2 replied:
@spraints Thanks. https://github.com/git-tfs/git-tfs/pull/1076

spraints commented:
> Can someone verify whether this means the test is wrong or the implementation is wrong? https://ci.appveyor.com/project/pmiossec/git-tfs-v2qcm/build/0.25.1.539#L609

I'd need to see what's in the branch's history and tree. can you add something to the smoke test that does `git log -p <brokenbranch>` and/or `git ls-tree -r <brokenbranch>`?

jnm2 commented on GitTfsTest/Integration/CloneTests.cs:
```diff
@@ -387,5 +388,45 @@ public void CloneWithAllBranchesShouldHandleFolderDeletedAndRecreatedAsBranch()
+                    .Change(TfsChangeType.Rename, TfsItemType.Folder, "$/MyTeamProject/BranchRenamed")
+                    .Change(TfsChangeType.Delete, TfsItemType.Folder, "$/MyTeamProject/Branch")
+                    .Change(TfsChangeType.Rename, TfsItemType.File, "$/MyTeamProject/BranchRenamed/File.txt", new byte[0], itemId: 100)
+                    .Change(TfsChangeType.Delete, TfsItemType.File, "$/MyTeamProject/Branch/File.txt");
```
Great, that makes more sense! I I had been copying this file via Find All References on `TfsChangeType.Rename`:
https://github.com/git-tfs/git-tfs/blob/master/GitTfsTest/Fixtures/vtccds.cs#L62-L64
https://github.com/git-tfs/git-tfs/blob/master/GitTfsTest/Fixtures/vtccds.cs#L49-L53
etc.

jnm2 commented:
When I removed the file delete change from the rename changeset it made no difference, but when I removed the branch folder delete, my test started failing again. I need to figure out why before I can look at the smoke tests again.

I want to stress that I don't have a good idea why my solution was working. The whole rename logic implementation is unclear to me.

jnm2 commented:
@spraints If I'm renaming an entire branch, does that need to be faked via `Changeset` or `BranchChangeset`?

jnm2 commented:
I think it has to be `BranchChangeset`. If I use just `Changeset` for the rename, it goes into a whole separate kind of infinite loop because `FindOrInitTfsRemoteOfChangeset` starts returning null for that changeset. Testing against my real TFS project, `FindOrInitTfsRemoteOfChangeset` does not return null, thereforce I'm going to use `BranchChangeset` as a test.

This also means I discovered a whole new type of infinite loop which probably needs its own test.

jnm2 commented:
@spraints In the real TFS project, the branch rename changeset that is tripping up git-tfs has only folder and file rename changes, but git-tfs is listing them all as deletes and `ChangeSieve.DeletesProject` is returning true. I'm worried that this isn't a minor glitch but that git-tfs is fundamentally misunderstanding something.

jnm2 commented:
Ah, they are all `Delete | SourceRename` in the TFS object model. But `git-tfs` thinks that manes `ChangeSieve.DeletesProject` applies and `ChangeSieve.DeletesProject` causes everything to short circuit and say there are no changes.

And incidentally, `IsBranch` is false for all items in the TFS object model. Which would seem to imply that I should not be using MergeChangeset in the fake setup.

jnm2 commented:
Ah, they are all `Delete | SourceRename` in the TFS object model. But `git-tfs` thinks that means `ChangeSieve.DeletesProject` applies and `ChangeSieve.DeletesProject` causes everything to short circuit and say there are no changes.

And incidentally, `IsBranch` is false for all items in the TFS object model. Which would seem to imply that I should not be using BranchChangeset for the rename changeset in the fake setup.

jnm2 commented:
I started over and got a fix for my real TFS project. Now I've been trying to find a way to get the faking system to simulate my real-world test. So far I can't get anything to match the before-fix and after-fix behavior of the real-world test.

pmiossec commented:
>Can someone verify whether this means the test is wrong or the implementation is wrong?

It seems to be the implementation that is wrong :(

I put the expected result in the repository: https://github.com/pmiossec/GitTfsSmokeTestResult 

You could `git clone --mirror https://github.com/pmiossec/GitTfsSmokeTestResult.git` and explore the bare repository to see what is wrong. Compare it to the repository cloned with your changes :
`git-tfs.exe clone https://tfs.codeplex.com:443/tfs/TFS16 $/vtccds/trunk --branches=all --username=vtccds_cp@snd --password=vtccds`

The problem seems to come from the branch `renamedTwice` (after commit `c7b539f089a0119f9c5370f2dae6276c83f369bc`) and the branch `renamed3` based on it --your change create even strange branches when trying to clone it :( --).

Every other branches are OK.

PS: sorry for all the branches with underscore prefix that is a mess. I really have to find time to fix that :(

jnm2 commented:
@pmiossec thanks, that helps! I'll try to turn the smoke test into an integration test.

pmiossec commented:
If you achieve that, ❤️!

pmiossec commented:
>PS: sorry for all the branches with underscore prefix that is a mess. I really have to find time to fix that :(

I have fixed it and merged to `master`, so if you think it could help you, you could rebase your work into onto last `master`...

jnm2 commented:
I ran out of time on this one and just now have confirmed that the problem is now fixed.
This was confirmed by comparing the details of https://github.com/git-tfs/git-tfs/pull/1077 to my memory of working on this. Thanks a ton @dolkensp for the fix and @spraints and @pmiossec for the help earlier!

jnm2 commented:
I do however, have a remaining issue. I can't get it to merge the three changesets leading up to a branch delete. Forcing it via `-c=8155`, `-c=8156`, and `-c=8157` just creates empty commits with the same message as changeset 8154. `-c=8158`, the branch delete, works properly but stops. I need this thing to keep going up to changeset 11926. =)

I'll probably start an issue when I have time. I am guessing I am yet again blocked still due to having to branch deleted and later a new branch created on top of it with teh exact same path, so TFS considers it a completely different branch but git-tfs confuses it with the first branch on the same path.

jnm2 commented:
I think this is it: https://github.com/git-tfs/git-tfs/issues/871
